### PR TITLE
[pipeline generator] Build out weekly test pipelines for all services and fix scheduling

### PR DESCRIPTION
**Pull request by benbp**
This PR updates the weekly test pipeline to run by default for all services (which will behave like our weekday live tests). It also fixes a scheduling/bucketing problem where the weekly tests were running on both Saturday and Sunday.

Update: this PR also adds name collision detection for pipeline definitions, with a helpful log message for a workaround. Resolves https://github.com/Azure/azure-sdk-tools/issues/2474
Resolves https://github.com/Azure/azure-sdk-tools/issues/2377

### CODE REVIEW

File: eng/pipelines/pipeline-generation.yml [hunk 1]
```diff
@@ -145,6 +145,12 @@ jobs:
     condition: ne(variables['TestOptions'],'')
     env:
       PATVAR: $(azuresdk-azure-sdk-devops-pipeline-generation-pat)
+  - script: |
+      $(Pipeline.Workspace)/pipeline-generator/pipeline-generator --organization https://dev.azure.com/azure-sdk --project internal --prefix $(Prefix) --path $(Pipeline.Workspace)/$(RepositoryName)/sdk --endpoint Azure --repository Azure/$(RepositoryName) --convention weekly --agentpool Hosted --branch refs/heads/$(DefaultBranch) --patvar PATVAR --debug $(TestOptions)
+    displayName: 'Generate weekly test pipelines (multi-cloud) for: $(RepositoryName)'
+    condition: ne(variables['TestOptions'],'')
```
Comment (weshaggard) [inline eng/pipelines/pipeline-generation.yml:151]:
```suggestion
    condition: and(succeeded(), ne(variables['TestOptions'],''))
```
Should probably fix the other condition above as well.

File: eng/common/pipelines/templates/steps/prepare-pipelines.yml [hunk 2]
```diff
@@ -74,6 +74,25 @@ steps:
...
+        --endpoint Azure
+        --repository ${{parameters.Repository}}
+        --convention weekly
+        --agentpool Hosted
+        --branch refs/heads/$(DefaultBranch)
+        --patvar PATVAR
+        --debug
+        ${{parameters.TestsConventionOptions}}
+      displayName: Create Weekly (Multi-Cloud) Live Test pipelines for public repository
+      condition: ne('${{parameters.TestsConventionOptions}}','')
```
Comment (weshaggard) [inline eng/common/pipelines/templates/steps/prepare-pipelines.yml:93]:
```suggestion
      condition: and(succeeded(), ne('${{parameters.TestsConventionOptions}}',''))
```

**Review by weshaggard — APPROVED** [review]
Couple comments/suggestions but otherwise looks good.

File: tools/pipeline-generator/Azure.Sdk.Tools.PipelineGenerator/Conventions/WeeklyIntegrationTestingPipelineConvention.cs [hunk 3]
```diff
@@ -23,10 +23,12 @@ protected override Schedule CreateScheduleFromDefinition(BuildDefinition definit
             var bucket = definition.Id % TotalBuckets;
             var startHours = bucket / BucketsPerHour;
             var startMinutes = bucket % BucketsPerHour;
+            var daysToBuild = new ScheduleDays[]{ ScheduleDays.Saturday, ScheduleDays.Sunday };
+            var dayBucket = definition.Id % daysToBuild.Length;
```
Comment (weshaggard) [inline tools/pipeline-generator/Azure.Sdk.Tools.PipelineGenerator/Conventions/WeeklyIntegrationTestingPipelineConvention.cs:27]: Hopefully we don't end up with a lot of even or odd id's which would cause more on one side or the other.

Comment (benbp) [reply tools/pipeline-generator/Azure.Sdk.Tools.PipelineGenerator/Conventions/WeeklyIntegrationTestingPipelineConvention.cs:27]: Currently there are 30 odd and 32 even definition IDs. If it gets super unbalanced we can figure something out or manually re-balance.

Comment (benbp) [reply tools/pipeline-generator/Azure.Sdk.Tools.PipelineGenerator/Conventions/WeeklyIntegrationTestingPipelineConvention.cs:27]: Though even if we have 1.5x as many on one day over another, there are still 96 15-minute buckets they get spread across, so I don't think the impact will be too bad compared to our normal weekday load.

File: eng/pipelines/pipeline-generation.yml (hunk 1 above)
Comment (benbp) [reply eng/pipelines/pipeline-generation.yml:151]: Do we want to fail fast, or try to create as many pipelines as we can (e.g. still create live pipelines if CI pipelines fail to create, etc.)?

Comment (weshaggard) [reply eng/pipelines/pipeline-generation.yml:151]: I think it is just good practice to stop if something fails or is canceled, in this case I think stopping is the right thing to do.

File: tools/pipeline-generator/Azure.Sdk.Tools.PipelineGenerator/Conventions/PipelineConvention.cs [hunk 4]
```diff
@@ -349,23 +349,19 @@ protected bool EnsureDefautPullRequestTrigger(BuildDefinition definition, bool o
...
-                        if (trigger.SettingsSourceType != 1 ||
-                            trigger.BranchFilters.Contains("+*"))
+                        // Override what is in the yaml file and use what is in the pipeline definition
+                        trigger.SettingsSourceType = 1;
+
+                        if (!trigger.BranchFilters.Contains("+*"))
                         {
-                            // Override what is in the yaml file and use what is in the pipeline definition
-                            trigger.SettingsSourceType = 1;
                             trigger.BranchFilters.Add("+*");
```
Comment (weshaggard) [inline tools/pipeline-generator/Azure.Sdk.Tools.PipelineGenerator/Conventions/PipelineConvention.cs:357]: It wasn't here before but do we need a hasChanges=true in this if block?

Comment (benbp) [reply tools/pipeline-generator/Azure.Sdk.Tools.PipelineGenerator/Conventions/PipelineConvention.cs:361]: Updated.

File: tools/pipeline-generator/Azure.Sdk.Tools.PipelineGenerator/Program.cs [hunk 5]
```diff
@@ -180,6 +184,12 @@ private PipelineConvention GetPipelineConvention(string convention, PipelineGene
                 }
 
                 logger.LogInformation("Found {0} components", components.Count());
+
+                if (HasPipelineDefinitionNameDuplicates(pipelineConvention, components))
```
Comment (weshaggard) [inline tools/pipeline-generator/Azure.Sdk.Tools.PipelineGenerator/Program.cs:188]: This might cause our pipeline generation to start failing so you need to be careful when you enable it for the first time.

File: tools/pipeline-generator/Azure.Sdk.Tools.PipelineGenerator/Program.cs [hunk 6]
```diff
@@ -236,5 +246,36 @@ private IEnumerable<SdkComponent> ScanForComponents(string path, string searchPa
...
         }
+
+        private bool HasPipelineDefinitionNameDuplicates(PipelineConvention convention, IEnumerable<SdkComponent> components)
+        {
+            var pipelineNames = new Dictionary<string, SdkComponent>();
+            var duplicates = new HashSet<SdkComponent>();
+
+            foreach (var component in components)
+            {
+                var definitionName = convention.GetDefinitionName(component);
```
Comment (weshaggard) [inline tools/pipeline-generator/Azure.Sdk.Tools.PipelineGenerator/Program.cs:257]: IIRC this causes a network call each time and we already do this in other places. Should we just add the dictionary building and checking there instead?

Comment (weshaggard) [reply tools/pipeline-generator/Azure.Sdk.Tools.PipelineGenerator/Program.cs:257]: Somewhere in CreateOrUpdateDefinitionAsync?

Comment (benbp) [reply tools/pipeline-generator/Azure.Sdk.Tools.PipelineGenerator/Program.cs:257]: `GetDefinitionName` is just a string building function, for example: https://github.com/Azure/azure-sdk-tools/blob/baefa12425b2ba30e5f60447655f1a295c81af3e/tools/pipeline-generator/Azure.Sdk.Tools.PipelineGenerator/Conventions/PullRequestValidationPipelineConvention.cs#L19

It's less efficient to do a separate check here as opposed to adding this logic incrementally in the `scan` step, but I think it's simpler this way and won't have any noticeable impact.

Comment (weshaggard) [reply tools/pipeline-generator/Azure.Sdk.Tools.PipelineGenerator/Program.cs:257]: OK as long as we are just processing things in memory and not making more network calls I'm cool with making it an explicit check.

**Review by weshaggard — APPROVED** [review]

File: tools/pipeline-generator/Azure.Sdk.Tools.PipelineGenerator/Program.cs (hunk 5 above)
Comment (benbp) [reply tools/pipeline-generator/Azure.Sdk.Tools.PipelineGenerator/Program.cs:188]: Tested locally, it passes for all repos (up, ci, tests, weekly conventions). Minor issue with local copies of node_modules causing duplicates, but that's an edge case that I think is ok and won't affect the pipeline.

File: tools/pipeline-generator/Azure.Sdk.Tools.PipelineGenerator/Program.cs (hunk 6 above)
Comment (benbp) [reply tools/pipeline-generator/Azure.Sdk.Tools.PipelineGenerator/Program.cs:257]: I also explicitly did it here vs. per update call, because I wanted to be able to collect all collisions in case there are more than 2, so that we can fail with full information about what needs to be fixed.
